### PR TITLE
boot-utils: Add support for booting Debian images

### PR DESCRIPTION
**Pull request by nathanchance**
Each of these commits should be self explanatory.

### CODE REVIEW

**Review by msfjarvis — APPROVED** [review]

**Comment by nathanchance** [comment]
I added support for `armhf`, `ppc64el`, and `s390x`. I have verified that they can all make it to userspace.

Additionally, I fixed an issue where `${INTERACTIVE}` was not getting a default value, which caused issues with using the script normally.

**Review by msfjarvis — APPROVED** [review]

File: README.txt [hunk 1]
```diff
@@ -30,6 +30,23 @@ Optional parameters:
...
     Invokes 'set -x' for debugging the script.
 
+  --debian:
+    By default, the script boots a very simple Busybox based root filesystem.
+    This option allows the script to boot a full Debian root filesystem,
+    which can be built using 'build.sh' in the debian folder. Run
+
+    $ sudo debian/build.sh -h
+
+    for more information that script.
```
Comment (nickdesaulniers) [inline README.txt:40]: grammar

File: boot-qemu.sh [hunk 2]
```diff
@@ -90,6 +94,11 @@ function sanity_check() {
     [[ -z ${ARCH} ]] && die "Architecture ('-a') is required but not specified!"
     [[ -z ${KERNEL_LOCATION} ]] && die "Kernel image or kernel build folder ('-k') is required but not specified!"
 
+    # --debian requires --interactive
+    [[ -z ${DEBIAN} ]] && DEBIAN=false
+    [[ -z ${INTERACTIVE} ]] && INTERACTIVE=false
+    ${DEBIAN} && ! ${INTERACTIVE} && die "'--debian' requires '-i'/'--interactive'!"
```
Comment (nickdesaulniers) [inline boot-qemu.sh:100]: how about the `--debian` flag just sets the `-i` flag implicitly?

File: boot-qemu.sh [hunk 3]
```diff
@@ -137,8 +155,11 @@ function setup_qemu_args() {
         arm32_v7)
             ARCH=arm
             APPEND_STRING+="console=ttyAMA0 "
+            # https://lists.nongnu.org/archive/html/qemu-discuss/2018-08/msg00030.html
+            # VFS: Cannot open root device "vda" or unknown-block(0,0): error -6
+            ${DEBIAN} && HIGHMEM=,highmem=off
```
Comment (nickdesaulniers) [inline boot-qemu.sh:160]: should the value assigned to `HIGHMEM` be a quoted string?

File: .gitignore [hunk 4]
```diff
@@ -1,2 +1 @@
 qemu-binaries/
-rootfs.cpio
```
Comment (nickdesaulniers) [inline .gitignore:2]: doesn't this mean then that each time we decompress an initrd, that the directory is now dirty?

File: debian/build.sh [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+# Prints a message in color
+function print_color() {
+    # Reset escape code
+    RST="\033[0m"
+    printf "\n%b%s%b\n" "${COLOR}" "${1}" "${RST}"
+}
+
+# Prints an error message in bold red then exits
+function die() {
+    COLOR="\033[01;31m" print_color "${1}"
```
Comment (nickdesaulniers) [inline debian/build.sh:14]: I'm surprised this works; can you just pass the color in as a normal parameter?

File: debian/build.sh [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+                    arm)
+                        DEB_ARCH=armhf
+                        OUR_ARCH=${1}
+                        ;;
+                    ppc64le)
+                        DEB_ARCH=ppc64el
+                        OUR_ARCH=${1}
+                        ;;
+                    s390)
+                        DEB_ARCH=${1}x
```
Comment (nickdesaulniers) [inline debian/build.sh:54]: for s390 you append `x`, for arm you spell out `armhf`. Let's be consistent here.

File: debian/build.sh [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+        flex
+        git
+        libtool
+        m4
+        pkg-config
+        stress-ng
+        sudo
+        vim
+    )
+    qemu-debootstrap --arch "${DEB_ARCH}" --include="${PACKAGES[*]//${IFS:0:1}/,}" buster "${MOUNT_DIR}" || exit ${?}
```
Comment (nickdesaulniers) [inline debian/build.sh:147]: do we want to move buster up to a global var near the top of the file, that way it's easier to upgrade to newer releases of debian some day?

File: debian/build.sh [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+
+cd "${LTP}" || exit ${?}
+
+make autotools
+./configure
+
+cd "${LTP}"/testcases/kernel/fs/proc && make -j$(nproc)
+cd "${LTP}"/testcases/kernel/fs/read_all && make -j$(nproc)
+cd "${LTP}"/testcases/lib && make -j$(nproc)
+EOF
```
Comment (nickdesaulniers) [inline debian/build.sh:176]: mind putting this in a separate file?

File: boot-qemu.sh [hunk 9]
```diff
@@ -285,7 +285,7 @@ function invoke_qemu() {
             # Note: no -serial mon:stdio
             "${QEMU[@]}" \
                 "${QEMU_ARCH_ARGS[@]}" \
-                -append "${APPEND_STRING}" \
+                -append "${APPEND_STRING%* }" \
```
Comment (nickdesaulniers) [inline boot-qemu.sh:332]: this is the kind of thing where in 6 months you look at and go "WTF is that?" The use of sigils is a perl-ism that I despise.

**Comment by nickdesaulniers** [comment]
Looks cool!

File: README.txt (hunk 1 above)
Comment (nathanchance) [reply README.txt:40]: Thanks, will change.

File: boot-qemu.sh (hunk 2 above)
Comment (nathanchance) [reply boot-qemu.sh:100]: Good catch, I forgot that we do that with `-g`/`--gdb`.

File: boot-qemu.sh (hunk 3 above)
Comment (nathanchance) [reply boot-qemu.sh:160]: `shellcheck` does not complain but I can update that if you feel it is worthwhile.

File: .gitignore (hunk 4 above)
Comment (nathanchance) [reply .gitignore:2]: No, I just moved it to `images/.gitignore`.

File: debian/build.sh (hunk 5 above)
Comment (nathanchance) [reply debian/build.sh:14]: Sure.

File: debian/build.sh (hunk 6 above)
Comment (nathanchance) [reply debian/build.sh:54]: Thanks, good catch.

File: debian/build.sh (hunk 7 above)
Comment (nathanchance) [reply debian/build.sh:147]: Yes and I could turn it into a parameter as well (`-v`/`--version` maybe?)

File: debian/build.sh (hunk 8 above)
Comment (nathanchance) [reply debian/build.sh:176]: Sure.

File: boot-qemu.sh (hunk 9 above)
Comment (nathanchance) [reply boot-qemu.sh:332]: Fair, I can pull it up and add a comment. Bash parameter expansion is better than shelling out to external utilities though so I would prefer not to eliminate it.

**Comment by nathanchance** [comment]
Thank you both for the review!

**Pull request merged**